### PR TITLE
Optimisation of historical forecast for regression models

### LINKED ISSUE #1233: historical_forecast enhancement for `RegressionModel`
**Is your feature request related to a current problem? Please describe.**
`RegressionModel` models type are specific models based on tabular data. There is a possibility to directly compute historical_forecast in the case **where retrain=False**, instead of using an iterative process for each time step. It can be particularly useful when some historical test set, after the train set chronologically, have to be computed to compute a test metric for example, and can save a lot of time.

**Describe proposed solution**
We will try to use or reaadapt the same tabularization function `create_lagged_data` and will create a matrix with lags etc to direclty compute the historical forecasts.

### PR DESCRIPTION
**Pull request by madtoinou**
Fixes #1233

### Summary

Reduce `historical_forecasts()` runtime for all the `RegressionModels` using some tricks (enforce `retrain=False`, use boundaries of the "forecastable indexes" instead of a range, predictions are vectorized with stride).

### Other Information

`forecast_horizon > model.output_chunk_length` is not supported at the moment because it would require auto-regression ~~and `num_samples > 1` needs a bit of work before being fully supported~~.

**[EDIT]** : Some speed up statistics, also included some comparison between the refactored and "legacy" implementation.

```python
# parameters
multi_models = [True, False]
multivariate = [True, False]
forecast_horizon = [1,7]
stride = 1
start = [700, 800, 900, 990]
length_ts = 1000

ts_univariate = tg.linear_timeseries(start_value=1, end_value=length_ts, length=length_ts)
ts_multivariate = ts_univariate.stack(tg.sine_timeseries(length=length_ts))

# two models, to test forecast_horizon different or equal to output_chunk_length
model1 = LinearRegressionModel(lags=3, output_chunk_length=forecast_horizon
model2 = LinearRegressionModel(lags=3, output_chunk_length=7)

# loop iterating on all the parameters
t0 = time.time()
for i in range(replicates):
    hist_fct = model.historical_forecasts(
       series=ts,
        start=start,
        stride=stride,
        last_points_only=last_points_only,
        forecast_horizon=forecast_horizon,
        num_samples=num_samples,
        retrain=False,
        enable_optimisation=False
        )
t1 = time.time()
for i in range(replicates):
    opti_hist_fct =  model.historical_forecasts(
        series=ts,
        start=start,
        stride=stride,
        last_points_only=last_points_only,
        forecast_horizon=forecast_horizon,
        num_samples=num_samples,
        retrain=False,
        enable_optimisation=True
t2 = time.time()
```

Refactorization of the original method did not affect the performance, but it should be easier to read.

The ratio on the y axis is `(t1-t0)/(t2-t1)`

### Speed up when `last_points_only=True`
legacy_v_opti
When the conditions are met, the gain scales with the length of the forecasted period.

### Speed up when `last_points_only=False`
legacy_v_opti_allpoints
The bottleneck is the creation of the returned `TimeSeries`, the gain are less significant.

### CODE REVIEW

File: darts/utils/optimised_historical_forecasts.py
Comment (dennisbader) [inline darts/utils/optimised_historical_forecasts.py:322]: can you explain why this is not required for ocl < forecast_horizon?

File: darts/utils/optimised_historical_forecasts.py
Comment (dennisbader) [inline darts/utils/optimised_historical_forecasts.py:356]: we can remove everything autoregressive related as we don't support it yet.
We can add this whenever we add support for it :)

File: darts/utils/optimised_historical_forecasts.py
Comment (dennisbader) [inline darts/utils/optimised_historical_forecasts.py:359]: this part is tricky to understand, could you add some more comments that describe how the output of `_predict_and_sample` is different for multi_model True/False, and share a bit more info for the slicing/reshaping.

We will be thankful in the future :)

File: darts/utils/optimised_historical_forecasts.py
Comment (dennisbader) [inline darts/utils/optimised_historical_forecasts.py:1]: This looks really great @madtoinou, congrats! This is an excellent boost historical forecasting 🚀 

Only suggestion is to give a bit more detail in some sections, as it can be tricky to wrap your head around all the different scenarios.

**Review by dennisbader — CHANGES_REQUESTED** [review]
Great PR, thanks a lot and congratulations @madtoinou! This will give such a boost to historical forecasting and backtesting for RegressionModels! 🚀 

I left some comments and suggestions regarding what we discussed offline, to add some more information on specific parts, etc.

File: darts/models/forecasting/forecasting_model.py
Comment (madtoinou) [reply darts/models/forecasting/forecasting_model.py:2030]: To prevent user from calling this method directly? Wouldn't be more efficient to just directly raise an error?

Aside from raising the error, what else would you expect this method to do? Because even if the sanity check pass, it's the sub-class `_optimized_historical_forecasts()` which is responsible for calling the appropriate optimized method.

File: darts/utils/optimised_historical_forecasts.py
Comment (madtoinou) [reply darts/utils/optimised_historical_forecasts.py:322]: it's the other way around, forecast_horizon < ocl, and it's because these timestamps will be covered by the last prediction of length `output_chunk_length`. By leaving them, the prediction will go too far.

File: darts/utils/optimised_historical_forecasts.py
Comment (madtoinou) [reply darts/utils/optimised_historical_forecasts.py:359]: Updated the comments, let me know if it's enough

File: darts/utils/optimised_historical_forecasts.py
Comment (madtoinou) [reply darts/utils/optimised_historical_forecasts.py:108]: Since `historical_forecasts_time_index` already corresponds to the "forecastable target" indexes, we could leave all the work to the tabularization module and pass the entire covariates series. I wanted to minimize the length of these series while making sure that tabularization had enough points to work with:
- push back the start of past covariates if some lags are defined
- push back the start of future covariates if some lags are negative
- push forward the end of future covariates if some lags are positive

I am going to push around the boundaries in all scenario, for extensiveness but I don't think that it'll make a big difference

File: darts/utils/optimised_historical_forecasts.py
Comment (madtoinou) [reply darts/utils/optimised_historical_forecasts.py]: I partially modularized this part, clearly separating these two aspects is quite difficult, let me know if it's enough!

File: darts/models/forecasting/regression_model.py
Comment (dennisbader) [inline darts/models/forecasting/regression_model.py:932]: in the current implemenetation models using encoders are not yet optimizable.
Do you think we could add support for this?

**Review by dennisbader — CHANGES_REQUESTED** [review]
Very, very cool stuff @madtoinou 🚀 

I really only had some minor suggestions :)
- add some tests
- add support for optimization with encoders (or at least make it non-optimizable if they use encoders)

We can

File: darts/models/forecasting/regression_model.py
Comment (madtoinou) [reply darts/models/forecasting/regression_model.py:860]: I try to favour Bristish spelling in generall but don't mind using US spelling in this context.

File: darts/models/forecasting/regression_model.py
Comment (madtoinou) [reply darts/models/forecasting/regression_model.py:932]: Good catch, I tend to forget about the encoders... Thank you for implementing this, I will adjust this PR as soon as the other one is merged.

File: darts/models/forecasting/forecasting_model.py
Comment (madtoinou) [reply darts/models/forecasting/forecasting_model.py:1993]: This method is never called for non-regression model, I wrote the warning just in case a user tries to call it directly.

Comment (dennisbader) [reply darts/models/forecasting/forecasting_model.py:2039]: True, then it's perfect :)

**Review by dennisbader — APPROVED** [review]
🚀 Super nice and great job @madtoinou 👏 
We're ready to merge now once unit tests have passed (i adapted slightly to reduce testing time).

**Pull request merged**
